File: reference/python/options.py
"""
Black-Scholes option pricing, Greeks, Newton-Raphson IV solver,
Monte Carlo with antithetic variates, and CRR binomial tree.

No external dependencies - uses math.erf for the normal CDF and
random.gauss for Monte Carlo draws.

All functions follow the same parameter convention:
  S           - spot price
  K           - strike price
  T           - time to expiry in years
  r           - continuous risk-free rate (e.g. 0.05 for 5%)
  sigma       - annualised volatility (e.g. 0.20 for 20%)
  option_type - 'call' or 'put'

Vega and Rho are scaled to a 1% move. Theta is per calendar day.
"""

import math
import random


def _norm_cdf(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def _norm_pdf(x: float) -> float:
    return math.exp(-0.5 * x * x) / math.sqrt(2.0 * math.pi)


def _d1_d2(S: float, K: float, T: float, r: float, sigma: float) -> tuple:
    d1 = (math.log(S / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * math.sqrt(T))
    d2 = d1 - sigma * math.sqrt(T)
    return d1, d2


def bs_price(S: float, K: float, T: float, r: float, sigma: float, option_type: str) -> float:
    d1, d2 = _d1_d2(S, K, T, r, sigma)
    disc = math.exp(-r * T)
    if option_type == "call":
        return S * _norm_cdf(d1) - K * disc * _norm_cdf(d2)
    return K * disc * _norm_cdf(-d2) - S * _norm_cdf(-d1)


def bs_greeks(S: float, K: float, T: float, r: float, sigma: float, option_type: str) -> dict:
    d1, d2 = _d1_d2(S, K, T, r, sigma)
    phi = _norm_pdf(d1)
    sqrt_T = math.sqrt(T)
    disc = math.exp(-r * T)

    delta = _norm_cdf(d1) if option_type == "call" else _norm_cdf(d1) - 1.0
    gamma = phi / (S * sigma * sqrt_T)
    vega = S * phi * sqrt_T / 100.0  # per 1% move in vol

    if option_type == "call":
        theta = (-(S * phi * sigma / (2.0 * sqrt_T)) - r * K * disc * _norm_cdf(d2)) / 365.0
        rho = K * T * disc * _norm_cdf(d2) / 100.0  # per 1% move in rate
    else:
        theta = (-(S * phi * sigma / (2.0 * sqrt_T)) + r * K * disc * _norm_cdf(-d2)) / 365.0
        rho = -K * T * disc * _norm_cdf(-d2) / 100.0

    return {"delta": delta, "gamma": gamma, "vega": vega, "theta": theta, "rho": rho}


def implied_vol(
    market_price: float,
    S: float,
    K: float,
    T: float,
    r: float,
    option_type: str,
    tol: float = 1e-7,
    max_iter: int = 100,
) -> float:
    """Newton-Raphson: iterate sigma until BS(sigma) = market_price."""
    sigma = 0.2
    for _ in range(max_iter):
        p = bs_price(S, K, T, r, sigma, option_type)
        d1, _ = _d1_d2(S, K, T, r, sigma)
        vega_raw = S * _norm_pdf(d1) * math.sqrt(T)  # unscaled, used as derivative
        diff = p - market_price
        if abs(diff) < tol:
            return sigma
        if abs(vega_raw) < 1e-12:
            break
        sigma -= diff / vega_raw
        sigma = max(1e-6, min(sigma, 10.0))  # keep sigma in a sane range
    return sigma


def monte_carlo(
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    option_type: str,
    n_paths: int = 100_000,
    seed: int = 42,
) -> dict:
    """
    Price via Monte Carlo with antithetic variates.

    Each standard normal draw Z is paired with -Z. For monotone payoffs,
    f(Z) and f(-Z) are negatively correlated, so averaging them reduces
    variance compared to two independent draws.
    """
    rng = random.Random(seed)
    half = n_paths // 2
    disc = math.exp(-r * T)
    drift = (r - 0.5 * sigma ** 2) * T
    vol_t = sigma * math.sqrt(T)

    payoffs = []
    for _ in range(half):
        z = rng.gauss(0.0, 1.0)
        for sign in (1.0, -1.0):
            ST = S * math.exp(drift + vol_t * sign * z)
            pf = max(ST - K, 0.0) if option_type == "call" else max(K - ST, 0.0)
            payoffs.append(disc * pf)

    n = len(payoffs)
    mean = sum(payoffs) / n
    variance = sum((p - mean) ** 2 for p in payoffs) / (n - 1)
    std_error = math.sqrt(variance / n)

    return {"price": mean, "std_error": std_error, "n_paths": n_paths}


def binomial_tree(
    S: float,
    K: float,
    T: float,
    r: float,
    sigma: float,
    option_type: str,
    n_steps: int = 200,
) -> dict:
    """
    CRR binomial tree returning both European and American prices.

    The early exercise premium (American - European) is most significant
    for deep ITM puts. For calls on a non-dividend-paying stock it is
    always zero (never optimal to exercise early).
    """
    dt = T / n_steps
    u = math.exp(sigma * math.sqrt(dt))
    d = 1.0 / u
    p = (math.exp(r * dt) - d) / (u - d)
    disc = math.exp(-r * dt)

    def payoff(s):
        return max(s - K, 0.0) if option_type == "call" else max(K - s, 0.0)

    # Terminal stock prices and payoffs
    V_eur = [payoff(S * u ** (n_steps - j) * d ** j) for j in range(n_steps + 1)]
    V_am = V_eur[:]

    # European: straightforward backward induction
    for _ in range(n_steps):
        V_eur = [disc * (p * V_eur[j] + (1.0 - p) * V_eur[j + 1]) for j in range(len(V_eur) - 1)]

    # American: same induction but check intrinsic at every node
    for i in range(n_steps - 1, -1, -1):
        V_am = [disc * (p * V_am[j] + (1.0 - p) * V_am[j + 1]) for j in range(i + 1)]
        for j in range(i + 1):
            intrinsic = payoff(S * u ** (i - j) * d ** j)
            V_am[j] = max(V_am[j], intrinsic)

    eur = V_eur[0]
    am = V_am[0]
    return {
        "european_price": eur,
        "american_price": am,
        "early_exercise_premium": am - eur,
    }

File: reference/r/options.R
# Black-Scholes option pricing, Greeks, Newton-Raphson IV solver,
# Monte Carlo with antithetic variates, and CRR binomial tree.
#
# No package dependencies - pnorm/dnorm/rnorm are all base R.
#
# Parameter conventions match the Python reference:
#   S           - spot price
#   K           - strike price
#   T           - time to expiry in years
#   r           - continuous risk-free rate (e.g. 0.05 for 5%)
#   sigma       - annualised volatility (e.g. 0.20 for 20%)
#   option_type - 'call' or 'put'
#
# Vega and Rho are scaled to a 1% move. Theta is per calendar day.

.d1_d2 <- function(S, K, T, r, sigma) {
    d1 <- (log(S / K) + (r + 0.5 * sigma^2) * T) / (sigma * sqrt(T))
    c(d1 = d1, d2 = d1 - sigma * sqrt(T))
}

bs_price <- function(S, K, T, r, sigma, option_type) {
    dd   <- .d1_d2(S, K, T, r, sigma)
    disc <- exp(-r * T)
    if (option_type == "call") {
        return(S * pnorm(dd["d1"]) - K * disc * pnorm(dd["d2"]))
    }
    K * disc * pnorm(-dd["d2"]) - S * pnorm(-dd["d1"])
}

bs_greeks <- function(S, K, T, r, sigma, option_type) {
    dd   <- .d1_d2(S, K, T, r, sigma)
    d1   <- dd["d1"]
    d2   <- dd["d2"]
    phi  <- dnorm(d1)
    sqT  <- sqrt(T)
    disc <- exp(-r * T)

    delta <- if (option_type == "call") pnorm(d1) else pnorm(d1) - 1.0
    gamma <- phi / (S * sigma * sqT)
    vega  <- S * phi * sqT / 100.0   # per 1% move in vol

    if (option_type == "call") {
        theta <- (-(S * phi * sigma / (2.0 * sqT)) - r * K * disc * pnorm(d2))  / 365.0
        rho   <- K * T * disc * pnorm(d2) / 100.0   # per 1% move in rate
    } else {
        theta <- (-(S * phi * sigma / (2.0 * sqT)) + r * K * disc * pnorm(-d2)) / 365.0
        rho   <- -K * T * disc * pnorm(-d2) / 100.0
    }

    list(delta = delta, gamma = gamma, vega = vega, theta = theta, rho = rho)
}

implied_vol <- function(market_price, S, K, T, r, option_type,
                        tol = 1e-7, max_iter = 100L) {
    sigma <- 0.2
    for (i in seq_len(max_iter)) {
        p        <- bs_price(S, K, T, r, sigma, option_type)
        d1       <- .d1_d2(S, K, T, r, sigma)["d1"]
        vega_raw <- S * dnorm(d1) * sqrt(T)  # unscaled, used as derivative
        diff     <- p - market_price
        if (abs(diff) < tol)       return(sigma)
        if (abs(vega_raw) < 1e-12) break
        sigma <- sigma - diff / vega_raw
        sigma <- max(1e-6, min(sigma, 10.0))   # keep sigma in a sane range
    }
    sigma
}

monte_carlo <- function(S, K, T, r, sigma, option_type,
                        n_paths = 100000L, seed = 42L) {
    # Antithetic variates: pair each Z draw with -Z.
    # For monotone payoffs, f(Z) and f(-Z) are negatively correlated,
    # so averaging them reduces variance vs. two independent draws.
    set.seed(seed)
    half   <- n_paths %/% 2L
    Z      <- rnorm(half)
    Z_full <- c(Z, -Z)

    disc <- exp(-r * T)
    ST   <- S * exp((r - 0.5 * sigma^2) * T + sigma * sqrt(T) * Z_full)
    pf   <- if (option_type == "call") pmax(ST - K, 0.0) else pmax(K - ST, 0.0)
    disc_pf <- disc * pf

    price     <- mean(disc_pf)
    std_error <- sd(disc_pf) / sqrt(n_paths)

    list(price = price, std_error = std_error, n_paths = n_paths)
}

binomial_tree <- function(S, K, T, r, sigma, option_type, n_steps = 200L) {
    dt   <- T / n_steps
    u    <- exp(sigma * sqrt(dt))
    d    <- 1.0 / u
    p    <- (exp(r * dt) - d) / (u - d)
    disc <- exp(-r * dt)

    j    <- 0L:n_steps
    ST   <- S * u^(n_steps - j) * d^j
    payoff_fn <- if (option_type == "call") {
        function(s) pmax(s - K, 0.0)
    } else {
        function(s) pmax(K - s, 0.0)
    }

    V_eur <- payoff_fn(ST)
    V_am  <- payoff_fn(ST)

    # European: straightforward backward induction
    for (step in seq_len(n_steps)) {
        n     <- length(V_eur)
        V_eur <- disc * (p * V_eur[-n] + (1.0 - p) * V_eur[-1L])
    }

    # American: same induction but check intrinsic at every node
    for (i in seq(n_steps - 1L, 0L)) {
        n    <- length(V_am)
        V_am <- disc * (p * V_am[-n] + (1.0 - p) * V_am[-1L])
        j_nodes  <- 0L:i
        ST_nodes <- S * u^(i - j_nodes) * d^j_nodes
        V_am <- pmax(V_am, payoff_fn(ST_nodes))
    }

    eur <- V_eur[[1L]]
    am  <- V_am[[1L]]
    list(
        european_price        = eur,
        american_price        = am,
        early_exercise_premium = am - eur
    )
}
